### DISCUSSION TITLE
How to add javafx support?

### DISCUSSION THREAD

**Discussion by znxuz** [mfussenegger/nvim-jdtls#169] [Q&A] [answered]
Could just not find any help regarding adding javafx to the language server.
env variable PATH_TO_FX is set to `/usr/lib/jvm/java-11-openjfx/lib`,
I've tried to add following lines to java-lsp.sh
`--module-path=$PATH_TO_FX \`
`--add-modules=ALL-MODULE-PATH \`
or
`--add-modules javafx.base,javafx.controls,javafx.graphics`,
at the top right below
`GRADLE_HOME=/usr/share/java/gradle /usr/lib/jvm/default-runtime/bin/java \`
argument as well as at the very bottom, but still there is no `import javafx.*`.
I've downloaded the packages via `pacman` for javafx:
```sh
java11-openjfx
java11-openjfx-src
java11-openjfx-doc
```
and also the lib directory is populated with all the necessary jar files including `javafx.base.jar, javafx.controls.jar`.
So what's the missing config that I am knowing to add?
Thanks in advance.

**Comment by mfussenegger** [comment, marked as answer]
> and also the lib directory is populated with all the necessary jar files including

What build system does the project use?
I got the imports working by adding the following to a `build.gradle`:

```
plugins {
    id 'application'
    id 'org.openjfx.javafxplugin' version '0.0.10'
}

javafx {
    version = "11"
    modules = [ 'javafx.controls' ]
}
```

**Reply by znxuz** [reply]
I'm still llearning Java and just tried with a sample non-project file without any build system. I am able to compile and run the sample program via command line with env variable `$PATH_TO_FX=/usr/lib/jvm/java-11-openjfx/lib`
`javac --module-path $PATH_TO_FX --add-modules=ALL-MODULE-PATH HelloWorld.java`
and
`java --module-path $PATH_TO_FX --add-modules=ALL-MODULE-PATH HelloWorld`

So all what's missing is the configuration for the lsp/jdtls, which I do not have any clue.
Does it mean that it is mandatory to set up a project & build system for the lsp to recognize and import javafx modules?

**Reply by mfussenegger** [reply]
> Does it mean that it is mandatory to set up a project & build system for the lsp to recognize and import javafx modules?

It is the easy way to get eclipse.jdt.ls to recognize your project dependencies.
I think you could specify the dependencies also manually within your jdtls `config`:

```lua
  config.settings = {
    java = {
      project = {
        referencedLibraries = {
          '/path/to/dependencyA.jar',
          '/path/to/dependencyB.jar',
        },
      }
    }
  }
```

But I've never used that myself.

**Reply by znxuz** [reply]
Thank you, I will try to write a new lua section inside vimrc containing this config for jdtls.

**Reply by Myzel394** [reply]
@zijian-x did this work for you? I tried it myself but unfortunately it didn't work.
